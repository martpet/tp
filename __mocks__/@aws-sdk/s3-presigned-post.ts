export const createPresignedPost = vi.fn().mockResolvedValue('dummyPresignedPost');
