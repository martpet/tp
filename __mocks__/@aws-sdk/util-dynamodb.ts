export const unmarshall = vi
  .fn()
  .mockName('unmarshall')
  .mockReturnValue('dummyUnmarshalledItems');
