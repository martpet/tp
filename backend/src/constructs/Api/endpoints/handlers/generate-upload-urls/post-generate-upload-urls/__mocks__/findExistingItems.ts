export const findExistingItems = vi
  .fn()
  .mockName('findExistingItems')
  .mockResolvedValue(['dummyFingerprint1']);
