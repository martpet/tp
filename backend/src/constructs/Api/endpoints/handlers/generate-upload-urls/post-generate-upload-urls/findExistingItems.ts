import {
  BatchGetCommand,
  DynamoDBClient,
  DynamoDBDocument,
  photosTableOptions,
  unmarshall,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocument.from(ddbClient);

export const findExistingItems = async (hashes: string[]) => {
  const { tableName, partitionKey } = photosTableOptions;
  const result: string[] = [];

  async function batchGet(keys: Record<string, any>[]) {
    const batchGetCommand = new BatchGetCommand({
      RequestItems: { [tableName]: { Keys: keys } },
    });

    const { Responses, UnprocessedKeys } = await ddbDocClient.send(batchGetCommand);

    if (Responses) {
      Responses[tableName].forEach(({ hash }) => result.push(hash));
    }

    if (UnprocessedKeys) {
      const nextKeys = UnprocessedKeys[tableName]?.Keys?.map((key) => unmarshall(key));

      if (nextKeys) {
        await batchGet(nextKeys);
      }
    }
  }

  const keys = hashes.map((hash) => ({ [partitionKey.name]: hash }));
  const maxItems = 100;
  const promises = [];

  for (let i = 0; i < keys.length; i += maxItems) {
    const chunk = keys.slice(i, i + maxItems);
    promises.push(batchGet(chunk));
  }

  await Promise.all(promises);

  return result;
};
