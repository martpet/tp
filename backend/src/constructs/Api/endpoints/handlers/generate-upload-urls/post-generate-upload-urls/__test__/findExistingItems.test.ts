import { BatchGetCommand, DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import { itResolves, itSendsAwsCommand, photosTableOptions } from 'lambda-layer';

import { findExistingItems } from '../findExistingItems';

vi.mock('@aws-sdk/util-dynamodb');

const ddbMock = mockClient(DynamoDBDocumentClient);

const args = [['hash1']] as Parameters<typeof findExistingItems>;

beforeEach(() => {
  ddbMock.reset();

  ddbMock.on(BatchGetCommand).resolves({
    Responses: {
      [photosTableOptions.tableName]: [
        { hash: 'existingHash1' },
        { hash: 'existingHash2' },
      ],
    },
  });
});

describe('findExistingItems', () => {
  itSendsAwsCommand(BatchGetCommand, ddbMock, findExistingItems, args);
  itResolves(findExistingItems, args);

  describe('when called with 101 items', async () => {
    const argsClone = structuredClone(args);
    argsClone[0] = Array(101);

    it('sends "BatchGetCommand" 2 times', async () => {
      await findExistingItems(...argsClone);
      expect(ddbMock.commandCalls(BatchGetCommand).length).toBe(2);
    });
  });

  describe('when "BatchGetCommand" response is missing "Responses"', () => {
    beforeEach(() => {
      ddbMock.on(BatchGetCommand).resolves({});
    });
    itResolves(findExistingItems, args);
  });

  describe('when "BatchGetCommand" response has "UnprocessedKeys"', () => {
    beforeEach(() => {
      ddbMock
        .on(BatchGetCommand)
        .resolvesOnce({
          Responses: {
            [photosTableOptions.tableName]: [{ hash: 'existingHash3' }],
          },
          UnprocessedKeys: {
            [photosTableOptions.tableName]: { Keys: [{ dummyUnprocessedKey: '' }] },
          },
        })
        .resolvesOnce({
          Responses: {
            [photosTableOptions.tableName]: [{ hash: 'existingHash4' }],
          },
        });
    });

    it('sends "BatchGetCommand" a second time with correct args', async () => {
      await findExistingItems(...args);
      expect(ddbMock.commandCalls(BatchGetCommand)[1].args[0].input).toMatchSnapshot();
    });

    itResolves(findExistingItems, args);
  });
});
