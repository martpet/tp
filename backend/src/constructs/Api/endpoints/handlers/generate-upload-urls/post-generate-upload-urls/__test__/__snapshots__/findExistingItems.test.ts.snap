// Vitest Snapshot v1

exports[`findExistingItems > resolves with a correct value 1`] = `
[
  "existingFingerprint1",
  "existingFingerprint1",
]
`;

exports[`findExistingItems > sends "BatchGetCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "RequestItems": {
    "TP-photos": {
      "Keys": [
        {
          "fingerprint": "fingerprint1",
        },
      ],
    },
  },
}
`;

exports[`findExistingItems > when "BatchGetCommand" response has "UnprocessedKeys" > resolves with a correct value 1`] = `
[
  "existingFingerprint3",
  "existingFingerprint4",
]
`;

exports[`findExistingItems > when "BatchGetCommand" response has "UnprocessedKeys" > sends "BatchGetCommand" a second time with correct args 1`] = `
{
  "RequestItems": {
    "TP-photos": {
      "Keys": [
        "dummyUnmarshalledItems",
      ],
    },
  },
}
`;

exports[`findExistingItems > when "BatchGetCommand" response is missing "Responses" > resolves with a correct value 1`] = `[]`;
